named_entity_recognition_bert/Legal-NER.ipynb

```diff
@@ -0,0 +1,1802 @@
+    "Let's now declare class for our token classification model and implement the training loop."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 26,
+   "id": "72d0f1ed",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "class BertModel(torch.nn.Module):\n",
+    "\n",
+    "    def __init__(self):\n",
+    "\n",
+    "        super(BertModel, self).__init__()\n",
+    "\n",
+    "        self.bert = BertForTokenClassification.from_pretrained('bert-base-cased', num_labels=len(unique_labels))\n",
+    "\n",
+    "    def forward(self, input_id, mask, label):\n",
+    "\n",
+    "        output = self.bert(input_ids=input_id, attention_mask=mask, labels=label, return_dict=False)\n",
+    "\n",
+    "        return output"
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 27,
+   "id": "b3c5e3ee",
+   "metadata": {
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "def train_loop(model, train_texts, train_entities, val_texts, val_entities, label2id):\n",
+    "\n",
+    "    train_dataset = DataSequence(train_texts, train_entities, label2id)\n",
+    "    val_dataset = DataSequence(val_texts, val_entities, label2id)\n",
+    "\n",
+    "    train_dataloader = torch.utils.data.DataLoader(train_dataset,\n",
+    "                                                   num_workers=4,\n",
+    "                                                   batch_size=BATCH_SIZE,\n",
+    "                                                   shuffle=True)\n",
+    "    \n",
+    "    val_dataloader = torch.utils.data.DataLoader(val_dataset,\n",
+    "                                                 num_workers=4,\n",
+    "                                                 batch_size=BATCH_SIZE)\n",
+    "    \n",
+    "    train_size = len(train_texts)\n",
+    "    val_size = len(val_texts)\n",
+    "\n",
+    "    use_cuda = torch.cuda.is_available()\n",
+    "    device = torch.device(\"cuda\" if use_cuda else \"cpu\")\n",
+    "    \n",
+    "    optimizer = NAdam(model.parameters(), lr=LEARNING_RATE)\n",
+    "\n",
+    "    if use_cuda:\n",
+    "        model = model.cuda()\n",
+    "\n",
+    "    best_acc = 0\n",
+    "    best_loss = 1000\n",
+    "\n",
+    "    for epoch_num in range(EPOCHS):\n",
+    "\n",
+    "        total_acc_train = 0\n",
+    "        total_loss_train = 0\n",
+    "\n",
+    "        model.train()\n",
+    "\n",
+    "        for train_data, train_label in tqdm(train_dataloader):\n",
+    "            train_label = train_label.to(device)\n",
+    "            mask = train_data['attention_mask'].squeeze(1).to(device)\n",
+    "            input_id = train_data['input_ids'].squeeze(1).to(device)\n",
+    "\n",
+    "            optimizer.zero_grad()\n",
+    "            loss, logits = model(input_id, mask, train_label)\n",
+    "\n",
+    "            for i in range(logits.shape[0]):\n",
+    "\n",
+    "                logits_clean = logits[i][train_label[i] != -100]\n",
```

Comment:
What is -100? Move it into some constant with explanation

Reply:
+